File: 0814-binary-tree-pruning/0814-binary-tree-pruning.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pruneTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def solve(root):
            if not root: return False
            left = solve(root.left)
            right = solve(root.right)
            if not left: 
                root.left = None
            if not right:
                root.right = None
            return root.val == 1 or left or right
        
        solve(root)
        if not root or (root.left == root.right == None and root.val == 0): return None
        return root

File: 0814-binary-tree-pruning/0814-binary-tree-pruning.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    bool pruneTreeModified(TreeNode* root) {
        if(!root) return false;
        
        bool leftTree = pruneTreeModified(root->left);
        bool rightTree = pruneTreeModified(root->right);
        
        if(!leftTree) {
            root->left = nullptr;
        }
        if(!rightTree) {
            root->right = nullptr;
        }

        return root->val == 1 || leftTree || rightTree;
    }
    
    TreeNode* pruneTree(TreeNode* root) {
    
        pruneTreeModified(root);
        if(!root->left && !root->right && !root->val) return nullptr;
        return root;
    }
};